File: 1592/main.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main()
    {
        uint n = uint.Parse(Console.ReadLine());
        List<uint> v = new List<uint>();
        
        for (uint i = 0; i < n; i++)
        {
            string[] parts = Console.ReadLine().Split(':');
            uint h = uint.Parse(parts[0]);
            uint m = uint.Parse(parts[1]);
            uint s = uint.Parse(parts[2]);
            uint seconds = (s + 60 * (m + 60 * h)) % (12 * 60 * 60);
            v.Add(seconds);
        }

        v.Sort();
        uint sum = 12 * 60 * 60 * n - (uint)v.Sum(x => (long)x);

        uint min = uint.MaxValue;
        uint mint = 0;
        uint prev = 0;

        for (int i = 0; i < v.Count; i++)
        {
            uint current = v[i];
            sum += n * (current - prev) - 12 * 60 * 60;
            if (sum < min)
            {
                min = sum;
                mint = current;
            }
            prev = current;
        }

        if (mint < 3600)
            mint += 3600 * 12;

        Console.WriteLine($"{mint / 3600}:{(mint / 60) % 60:D2}:{mint % 60:D2}");
    }
}

File: 1592/main.py
def main():
    n = int(input())
    v = []

    for _ in range(n):
        h, m, s = map(int, input().split(':'))
        v.append((s + 60 * (m + 60 * h)) % (12 * 60 * 60))

    v.sort()
    total_seconds = 12 * 60 * 60 * n - sum(v)

    min_time = float('inf')
    mint = 0
    prev = 0

    for i in range(n):
        current = v[i]
        total_seconds += n * (current - prev) - 12 * 60 * 60
        if total_seconds < min_time:
            min_time = total_seconds
            mint = current
        prev = current

    if mint < 3600:
        mint += 3600 * 12

    print(f"{mint // 3600}:{(mint // 60) % 60:02}:{mint % 60:02}")

if __name__ == "__main__":
    main()
